File: Algorithms/127. Word Ladder/solution.js
/**
 * @param {string} beginWord
 * @param {string} endWord
 * @param {string[]} wordList
 * @return {number}
 */
var ladderLength = function (beginWord, endWord, wordList) {
  const unvisited = new Set(wordList);
  if (!unvisited.has(endWord)) return 0;
  if (beginWord === endWord) return 1;
  const queue = [beginWord];
  const alphabet = 'abcdefghijklmnopqrstuvwxyz';
  let length = 1;
  while (queue.length) {
    const queueLength = queue.length;
    for (let j = 0; j < queueLength; j++) {
      const word = queue.shift();
      for (let i = 0; i < word.length; i++) {
        for (const ch of alphabet) {
          const newWord = word.slice(0, i) + ch + word.slice(i + 1);
          if (unvisited.has(newWord)) {
            if (newWord === endWord) return length + 1;
            unvisited.delete(newWord);
            queue.push(newWord);
          }
        }
      }
    }
    length++;
  }
  return 0;
};

File: Algorithms/127. Word Ladder/solution.go
package problem127

func ladderLength(beginWord string, endWord string, wordList []string) int {
	if beginWord == endWord {
		return 1
	}

	wordMap := map[string]bool{}
	for _, word := range wordList {
		wordMap[word] = true
	}
	if !wordMap[endWord] {
		return 0
	}

	src := map[string]bool{beginWord: true}
	dist := map[string]bool{endWord: true}
	alphabet := "abcdefghijklmnopqrstuvwxyz"
	length := 2
	for len(src) > 0 && len(dist) > 0 {
		if len(src) > len(dist) {
			src, dist = dist, src
		}
		next := map[string]bool{}
		for word := range src {
			delete(wordMap, word)
			for i := 0; i < len(word); i++ {
				for _, ch := range alphabet {
					if byte(ch) == word[i] {
						continue
					}
					newWord := word[0:i] + string(ch) + word[i+1:]
					if dist[newWord] {
						return length
					}
					if wordMap[newWord] {
						delete(wordMap, newWord)
						next[newWord] = true
					}
				}
			}
		}
		length++
		src = next
	}
	return 0
}
